Fix ASAN issues

thesiv opened the pull request:
Fix the issues found in the GUI unit tests (`./test_gui`) rebuilt with `-fsanitize=address -fno-omit-frame-pointer`.

vadz committed:
Add a suppression file for FontConfig leaks occurring under wxGTK

We're not interested in these (one off) leaks and can't do anything
about them anyhow, so just provide a way too suppress leak sanitizer
errors about them.

thesiv committed:
Fix possible use of dangling window pointer in wxSizer::Insert()

Add the sizer item to the sizers items list only after calling of
wxWindowBase::SetContainingSizer() because it can throw (if it asserts
and the assert handler throws an exception, as happens in our own unit
tests) and then the sizer item would be kept in the sizers items list
but m_containingSizer wouldn't be set for the window.

thesiv committed:
Fix wxAcceleratorEntry memory leaks in the menu test

Use wxScopedPtr to ensure that all the test accelerator entries are
destroyed instead of just being leaked.

thesiv committed:
Fix memory leaks in wxHtmlParser unit test

Remove the unnecessary, and actually harmful, overridden GetProduct(),
as we must delete the pointer returned by the base class version to
avoid leaking it.

Also use wxScopedPtr to avoid leaking a test wxHtmlContainerCell.

thesiv committed:
Fix memory leak in wxGridSizer::Insert() if an assert in it fails

This avoids another memory leak in the test suite, which intentionally
triggers an assert in this function.

vadz commented on tests/html/htmlparser.cpp:
```diff
@@ -30,7 +30,13 @@ TEST_CASE("wxHtmlParser::ParseInvalid", "[html][parser][error]")
     {
     public:
-        virtual wxObject *GetProduct() wxOVERRIDE { return NULL; }
+        virtual wxObject *GetProduct() wxOVERRIDE
+        {
+            // Return the top level HTML container cell, which will be returned
+            // by wxHtmlWinParser::Parse(). And we will delete it to avoid the
+            // memory leaking.
+            return wxHtmlWinParser::GetProduct();
+        }
```
Why do we need this at all? I.e. why override it in the first place?

vadz commented on src/common/sizer.cpp:
```diff
@@ -676,15 +677,34 @@ wxSizer::~wxSizer()
-    m_children.Insert( index, item );
+    // Use the scoped pointer to avoid memory leaking when
+    // wxWindowBase::SetContainingSizer() throws.
+    wxScopedPtr<wxSizerItem> scoped_item( item );
```
Let's follow the usual naming conventions
```suggestion
    wxScopedPtr<wxSizerItem> scopedItem( item );
```

vadz commented on src/common/sizer.cpp:
```diff
@@ -1452,6 +1472,9 @@ wxGridSizer::wxGridSizer( int rows, int cols, const wxSize& gap )
 
 wxSizerItem *wxGridSizer::DoInsert(size_t index, wxSizerItem *item)
 {
+    // Ensure that the item will be deleted in case of exceptions throwing.
```
```suggestion
    // Ensure that the item will be deleted in case of exception.
```

vadz commented on tests/html/htmlparser.cpp:
```diff
@@ -104,6 +110,8 @@ TEST_CASE("wxHtmlCell::Detach", "[html][cell]")
             top->Detach(cell1);
         );
     }
+
+    delete top;
```
Could we use scoped pointer for `top` instead?

vadz commented on src/common/sizer.cpp:
```diff
@@ -676,15 +677,34 @@ wxSizer::~wxSizer()
+        }
+        catch ( ... )
+        {
+            item->DetachWindow();
```
Sorry, I don't understand why is this necessary. Without this call, and try/catch around it, if `SetContainingSizer()` throws, the scoped pointer containing `item` will delete it, and call `SetContainingSizer(NULL)` on the window -- but what's wrong with that?

vadz reviewed (commented):
Thanks, the other fixes look good, but I don't understand the one for `wxSizer::DoInsert()`.

Also, please add a build using ASAN to `.travis.yml` (or, maybe even better, change one of the existing builds to use it). TIA!

thesiv committed:
Add wxGTK build with ASAN to Travis CI

Compile and link with `-fsanitize=address` options to find memory related
errors.

vadz commented on src/common/sizer.cpp:
```diff
@@ -676,15 +677,45 @@ wxSizer::~wxSizer()
-    m_children.Insert( index, item );
+    // The helper class that solves two problems when
+    // wxWindowBase::SetContainingSizer() throws:
+    // 1. Avoid the memory leaking using the scoped pointer to the sizer item.
```
```suggestion
    // 1. Avoid leaking memory using the scoped pointer to the sizer item.
```
(or "Avoid the memory leaks")

vadz commented on src/common/sizer.cpp:
```diff
@@ -676,15 +677,45 @@ wxSizer::~wxSizer()
+                m_item->DetachWindow();
+        }
+
+        wxSizerItem* Dismiss()
```
I know that I suggested using `Dismiss()` myself, but I thought it would be void -- as "dismiss" is an action without any return value. So I'd either make it void and then `return item` below or rename it to `Release()` -- whichever you prefer.

vadz reviewed (commented):
Thanks, this looks good, but the build fails due to some apparently real memory leaks -- could you please look into those? TIA!

thesiv committed:
Fix wxMemoryFSHandler memory leaks in wxFileSystem unit test

Use wxScopedPtr<> to ensure the handlers we use get freed.

thesiv committed:
Fix list contents memory leak in wxAny unit test

Also document the need for using WX_CLEAR_LIST() when converting a
list-valued wxVariant to wxAny.

Note that we intentionally don't fix the problem by clearing the list
automatically, even if it could be done, because this would silently
break the existing code which does already clear the list -- and now
would attempt to clear it twice, with fatal consequences. Instead
document the existing behaviour and explain how to avoid memory leaks.

thesiv committed:
Suppress spurious accessibility bus errors in Travis CI builds

These errors don't matter, so just avoid them in the build logs by
setting NO_AT_BRIDGE environment variable.

vadz commented on build/tools/before_install.sh:
```diff
@@ -42,6 +42,23 @@ case $(uname -s) in
+
+            if [ "$wxUSE_ASAN" = 1 ]; then
+                # Enable the `-dbgsym` repositories.
+                echo "deb http://ddebs.ubuntu.com $(lsb_release -cs) main restricted universe multiverse
```
It is better to use long options in the scripts, especially if they're not well-known (e.g. `rm -f` is probably fine). So could you please do
```suggestion
                echo "deb http://ddebs.ubuntu.com $(lsb_release --codename --short) main restricted universe multiverse
```
and the same below (alternatively define a variable `$codename` once and use it)?

vadz commented on .travis.yml:
```diff
@@ -10,96 +10,10 @@ dist: xenial
 
 jobs:
     include:
-        - dist: trusty
```
Just a reminder, but this needs to be reverted.

vadz commented on build/tools/travis-ci.sh:
```diff
@@ -17,6 +20,11 @@ case $wxTOOLSET in
         cmake --version
 
+        if [ "$wxUSE_ASAN" = 1 ]; then
+            echo "ASAN currently don't supported for CMAKE builds"
```
```suggestion
            echo "ASAN currently isn't supported in CMake builds"
```

vadz commented on build/tools/travis-ci.sh:
```diff
@@ -129,7 +150,12 @@ case $wxTOOLSET in
             echo 'Testing GUI using Xvfb...'
             pushd tests
-            xvfb-run -a -s '-screen 0 1600x1200x24' ./test_gui
+            export tests="RectTestCase SizeTestCase PointTestCase RealPointTestCase RegionTestCase BitmapTestCase::Mask BitmapTestCase::OverlappingBlit BitmapTestCase::DrawNonAlphaWithMask BitmapTestCase::DrawAlpha BitmapTestCase::DrawAlphaWithMask BitmapTestCase::SubBitmapNonAlpha BitmapTestCase::SubBitmapNonAlphaWithMask BitmapTestCase::SubBitmapAlphaWithMask DC::Clear GC::DrawBitmap wxColour:: ...
```
OK, this definitely merits an explanation -- why do we need to run each test individually?

Also, even if this is really needed (and I hope we can avoid it), the list of tests must be retrieved from `./test_gui --list-test-names-only` output and not hardcoded like this, as this is completely unmaintainable.

vadz commented on interface/wx/any.h:
```diff
@@ -84,6 +84,24 @@ class wxAny
                  hold and manage reference to wxVariantData* similar to how
                  wxVariant does.
+
+                 Note that constructing from the list variant produces the
+                 memory leak if the list don't cleared explicitly. Don't fixed
+                 to not break the existing code.
```
```suggestion
                 Note that objects constructed from list-valued variants
                 require the list to be explicitly cleared using `WX_CLEAR_LIST`
                 to avoid leaking memory. This unfortunate behaviour will not
                 be changed to prevent breaking the existing code relying on it.
```

vadz reviewed (commented):
Thanks for fixing all these issues!

The changes to the code look fine to me, but there are still some questions about the scripts, please see below.

thesiv committed:
Fix memory leak in wxWebKitJavascriptResult

We need to unref the result wrapped by this object in its dtor.

thesiv committed:
Fix global read overrun error in MBConvTestCase

Use null-terminated copy of the input buffer to create the formatted
message instead of "multiBuffer" itself, which isn't.

vadz committed:
Remove useless --disable-optimise option from ANSI Travis build

This is already added unconditionally in build/tools/travis-ci.sh
anyhow, so there is no need to specify it here.

vadz commented:
Thanks, look good to me, I've just rebased everything on master and adjusted some commit messages and will merge it now once the CI builds pass.

vadz committed:
Specify wxGTK_VERSION explicitly for all Travis CI builds

It default to 2 currently, which is rather surprising, so prepare for
requiring specifying it explicitly by adding wxGTK_VERSION to all wxGTK
builds environments.

vadz committed:
Require specifying wxGTK_VERSION for wxGTK builds on Travis CI

This avoids accidentally using a wrong GTK version, as it happened for
e.g. ARM64 build.

vadz committed:
Select GTK version explicitly when configuring Travis CI builds

Don't rely on just a single GTK version being installed on the system,
as we do now, but select the version we want explicitly, just to avoid
bad surprises in the future (even if it doesn't change anything for
now).
